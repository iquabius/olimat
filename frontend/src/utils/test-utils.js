import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from 'react-testing-library';
import { MockedProvider } from 'react-apollo/test-utils';

// eslint-disable-next-line import/prefer-default-export
export const renderApollo = (
  node,
  { mocks, addTypename = false, defaultOptions, cache, ...options } = {},
) => {
  return render(
    <MockedProvider
      mocks={mocks}
      addTypename={addTypename}
      defaultOptions={defaultOptions}
      cache={cache}
    >
      {node}
    </MockedProvider>,
    options,
  );
};
