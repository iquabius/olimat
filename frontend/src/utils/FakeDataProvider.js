// Este arquivo só é usado no servidor, pra testes.
// Podemos importar dependências de desenvolvimento.
// Talvez não funcione num ambiente de Integração Contínua (CI).
/* eslint import/no-extraneous-dependencies: ["error", {"devDependencies": true}] */
import React from 'react';
import PropTypes from 'prop-types';
import { addMockFunctionsToSchema } from 'graphql-tools';
import ApolloClient from 'apollo-client';
import { SchemaLink } from 'apollo-link-schema';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloProvider } from 'react-apollo';
import { globalMocks, schema } from './test-utils';
import mergeResolvers from './mergeResolvers';

// globalMocks uses fake data generated by faker.js
// It is merged with customResolvers, which can be used to customize the data
// returned by specific fields.
const FakeDataProvider = props => {
  const { children, customResolvers = {} } = props;
  const mocks = mergeResolvers(globalMocks, customResolvers);
  addMockFunctionsToSchema({ schema, mocks });

  const client = new ApolloClient({
    link: new SchemaLink({ schema }),
    cache: new InMemoryCache(),
  });

  return <ApolloProvider client={client}>{children}</ApolloProvider>;
};

FakeDataProvider.propTypes = {
  /**
   * Used to render any component that uses a <Query/> or <Mutation/>
   * from 'react-apollo'.
   * The component is wrapped by <ApolloProvider/>, with a mocked ApolloClient.
   */
  children: PropTypes.node.isRequired,
  /**
   * An object with resolvers to be merged with the fake data resolvers.
   * It can be used to override fake data of specific fields:
   * customResolvers = {
   *   Question: () => ({
   *     wording: 'Maria comprou três maças...',
   *     secondaryWording: null,
   *     imageFullUrl: 'http://host.test/image.jpg',
   *     choices: [{ text: '5 maças' }, { text: '3 maças' }],
   *   }),
   * }
   */
  customResolvers: PropTypes.object,
};

export default FakeDataProvider;
