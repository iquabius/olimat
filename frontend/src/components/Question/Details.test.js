/* eslint-env jest */
import React from 'react';
import QuestionDetails from './Details';
import { renderApollo } from '../../utils/test/test-utils';
import { waitForElement, fireEvent, render } from 'react-testing-library';
import Router from 'next/router';
import { SnackbarProvider } from 'notistack';
import FakeDataProvider from '../../utils/test/FakeDataProvider';
import MockErrorProvider from '../../utils/test/MockErrorProvider';
import MockNextContext from '../../utils/test/MockNextContext';

const MockQuestionDetails = () => (
  <MockNextContext router={{ query: { id: 'theQuestionId1' } }}>
    <SnackbarProvider maxSnack={3}>
      <QuestionDetails />
    </SnackbarProvider>
  </MockNextContext>
);

describe('<QuestionDetails />', () => {
  test('renders loading state initially', () => {
    const { getByText } = renderApollo(<MockQuestionDetails />);
    getByText(/carregando/i);
  });

  test('renders error message', async () => {
    const errorMsg = 'Que pena';

    const { getByText } = render(
      <MockErrorProvider graphqlErrors={[{ message: errorMsg }]}>
        <MockQuestionDetails />
      </MockErrorProvider>,
    );

    await waitForElement(() => getByText(errorMsg, { exact: false }));
  });

  test('renders the details of a question', async () => {
    const customResolvers = {
      Question: () => ({
        wording: 'Maria comprou três maças...',
        secondaryWording: null,
        imageFullUrl: 'http://host.test/image.jpg',
        choices: [{ text: '5 maças' }, { text: '3 maças' }],
      }),
    };
    const { container, getByText, getByAltText } = render(
      <FakeDataProvider customResolvers={customResolvers}>
        <MockQuestionDetails />
      </FakeDataProvider>,
    );

    await waitForElement(() => getByText(customResolvers.Question().wording));

    // TODO: Add imageAltText to Question type
    const questionImage = getByAltText('Imagem da questão');
    expect(questionImage).toHaveAttribute('src', customResolvers.Question().imageFullUrl);

    // Confere se a lista de alternativas foi renderizada,
    // com a quantidade certa de itens
    const choicesList = container.querySelector('ul');
    expect(choicesList).toHaveTextContent(customResolvers.Question().choices[0].text);
    expect(choicesList.children.length).toBe(customResolvers.Question().choices.length);
  });

  test('clicking on delete button shows the warning dialog', async () => {
    const { getByText, getByLabelText } = render(
      <FakeDataProvider>
        <MockQuestionDetails />
      </FakeDataProvider>,
    );

    await waitForElement(() => getByLabelText('Excluir questão'));

    const deleteButton = getByLabelText('Excluir questão');
    fireEvent.click(deleteButton);

    expect(deleteButton).toBeDisabled();
    getByText(/será apagada permanentemente/);
  });

  test('clicking on cancel button hides the warning dialog', async () => {
    const { getByLabelText } = render(
      <FakeDataProvider>
        <MockQuestionDetails />
      </FakeDataProvider>,
    );

    await waitForElement(() => getByLabelText('Excluir questão'));

    const deleteButton = getByLabelText('Excluir questão');
    fireEvent.click(deleteButton);

    expect(deleteButton).toBeDisabled();

    const cancelDeleteButton = getByLabelText('Cancelar excluir questão');
    fireEvent.click(cancelDeleteButton);

    expect(deleteButton).not.toBeDisabled();
  });

  test('clicking edit link should go to edit form page', async () => {
    const customResolvers = {
      Question: () => ({
        // Used by the router to send users to edit question page
        id: 'AranDOMquestionID',
      }),
    };
    const { getByLabelText } = render(
      <FakeDataProvider customResolvers={customResolvers}>
        <MockQuestionDetails />
      </FakeDataProvider>,
    );

    await waitForElement(() => getByLabelText('Editar questão'));

    const editLink = getByLabelText('Editar questão');
    fireEvent.click(editLink);

    expect(Router.pathname).toMatchSnapshot();
  });
});
