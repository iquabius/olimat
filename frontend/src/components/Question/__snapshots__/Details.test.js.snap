// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QuestionDetails /> clicking edit link should go to edit form page 1`] = `"/admin/questao-editar?id=AranDOMquestionID"`;

exports[`<QuestionDetails /> clicking on confirm button deletes the question 1`] = `"/admin/questoes"`;
